src/server/cluster/outgoing_slot_migration.cc

```diff
@@ -112,6 +113,7 @@ void OutgoingMigration::Finish(bool is_error) {
     return;
   }
 
+  slot_migrations_init_guard_.Wait();
```

Comment:
intuitively I find it not right that you wait for all the callbacks to finish in one of the callbacks. Moreover, seems from the first sight  redundant since you already use  AwaitFiberOnAll as a barrier.

Reply:
> intuitively I find it not right that you wait for all the callbacks to finish in one of the callbacks

Why, though? That callback can yield to other callbacks if needed. It's not blocking on the shard queue, so what's wrong with that?
(further, one could argue that basically everything that runs in Dragonfly is callbacks :) )

> Moreover, seems from the first sight redundant since you already use AwaitFiberOnAll as a barrier.

That's what I had in mind when I fixed this code originally (a few days ago). But the problem is that cancellation can happen in another thread (it comes from the client via the admin-port). So `AwaitFiberOnAll` wouldn't do on its own :(

Reply:
Callbacks can block, I am not saying there is a technical problem, but usually we do not synchronise callbacks between themselves when we run AwaitFiberOnAll - we wait for them to return all the information, aggregate it and then do the next step if needed. Maybe what's missing for me is to understand what the problematic use-case is.

Reply:
I think now I understand it actually. I would separate it into two 
`shard_set->pool()->Await(Fiber?)OnAll` steps: the first one initializing the migration, and the second one doing the cancellable sync